Improve cprover_string_literal argument handling TG-1245

romainbrenguier opened the pull request:
cprover_string_literal was requiring a string constant as argument,
but we now deal with if expressions as well.
This also improve solver handling of expressions of type `string_typet` by recording the symbols in the symbol_resolve map.

romainbrenguier committed:
Add an optional guard to add_axioms_for_constant

This allows to add a condition under which we want the result to be
equal to the given constant.
By default this is true, meaning that the result should always equal the constant.

romainbrenguier committed:
Handle if_exprt in add_axioms_for_string_literal

cprover_string_literal was requiring a string constant as argument,
but we now deal with if expressions as well.
So expressions like:
`cprover_string_literal_func!0#0(cprover_string_length$3!0@1#2,
nondet_infinite_array$5!0@1, this!0@1#1 == &tmp_object_factory$17!0 ?
tmp_object_factory$17!0#4.@class_identifier :
this$object#0.@class_identifier)`
are correctly handled by the string solver.

smowton commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -336,6 +336,32 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
+    }
+  }
+
+  for(const auto &t : type.subtypes())
```
This doesn't mean subtype in the Java sense, but rather the arguments of a type in the algebraic datatype sense -- for example, the "subtypes" of a struct are the types of its components (therefore the struct special case above is redundant), the "subtype" of a pointer is the type it points to... and so on.

romainbrenguier replied:
@smowton 
> therefore the struct special case above is redundant

I'm not sure it is: in a struct the subtypes appear in the components, which is a `named_sub` for the `irept` and not in `sub` which is what `subtypes` returns

romainbrenguier replied:
I added to the documentation the precision that we mean subtype in the algebraic sense

smowton replied:
Ah so it is. No problem then.

smowton reviewed (commented):
Not sure if `has_string_subtype` is doing what you want to, could you clarify the intent, perhaps by giving some examples of types you expect to pass and fail `has_string_subtype`'s test?

mgudemann commented on src/solvers/refinement/string_constraint_generator_constants.cpp:
```diff
@@ -65,6 +67,39 @@ exprt string_constraint_generatort::add_axioms_for_empty_string(
+/// \param res: string expression for the result
+/// \param arg: expression of type string typet
+/// \param guard: condition under which `res` should be equal to arg
+/// \return 0 if constraints were added, 1 if expression could not be handled
```
what is with the return value of `if_exprt` ? This reads as if only 1 or 0 can be returned.

romainbrenguier replied:
Yes, in case of if_expr we return the maximum of the true case or false case, so it will always be 0 or 1

mgudemann commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -336,6 +340,36 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
+  {
+    if(has_string_subtype(t, ns))
+      return true;
+  }
```
would for the above two loops using `std::any_of` be easier to understand?

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -312,7 +312,11 @@ bool is_char_pointer_type(const typet &type)
 /// \param ns: name space
-/// \return true if a subtype is an pointer of characters
+/// \return true if a subtype of `type` is an pointer of characters.
+///         The meaning of "subtype" is in the algebraic datatype sense:
```
May be being dense, but I don't completely get this, perhaps an example (e.g. a pointer to a char or struct with a char component)

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -336,6 +340,39 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
 }
 
+/// \param type: a type
+/// \param ns: name space
```
Nit: no space between name and space

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -336,6 +340,39 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
+///         etc...
+static bool has_string_subtype(const typet &type, const namespacet &ns)
+{
+  if(type == string_typet())
```
Can you pull the common code between this method and `has_char_pointer_subtype` into a separate method (`has_type_have_subtype(const typet & type, const namespacet &ns, const typet &required_subtype)`)

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -336,6 +340,39 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
+///         for example, the subtypes of a struct are the types of its
+///         components, the subtype of a pointer is the type it points to,
+///         etc...
+static bool has_string_subtype(const typet &type, const namespacet &ns)
```
It would be good to add a couple of unit tests for this if you don't mind.

thk123 commented on src/solvers/refinement/string_constraint_generator_constants.cpp:
```diff
@@ -67,6 +67,39 @@ exprt string_constraint_generatort::add_axioms_for_empty_string(
+  }
+  else if(const auto constant_expr = expr_try_dynamic_cast<constant_exprt>(arg))
+    return add_axioms_for_constant(res, constant_expr->get_value(), guard);
+  else
```
A comment explaining why in this case a `1` expression is a reasonable return might be good.

thk123 commented on src/solvers/refinement/string_constraint_generator_constants.cpp:
```diff
@@ -81,7 +114,5 @@ exprt string_constraint_generatort::add_axioms_from_literal(
-  const exprt &arg = args[2];
-  irep_idt sval=to_constant_expr(arg).get_value();
-  return add_axioms_for_constant(res, sval);
+  return add_axioms_for_cprover_string(res, args[2], true_exprt());
```
It would be good to see a test for whatever case is now supported. I can't say the changes in this file mean much to me

romainbrenguier replied:
There will be tests in other repositories: https://github.com/diffblue/test-gen/pull/1327
But it is difficule to test on CBMC because we don't have a model for Class and the main affected method is Object.getClass.

An example that should work afterwards is the following:
```java
public class testMaybeNull1
{
    public Object x;
    public String check()
    {
        return x.getClass().getName();
    }
}
```

The reason is that getClass converts a cprover_string to a java string (the `@class_identifier`), but since the object could be null, the argument passed to `cprover_string_literal` is more complicated than a constant.

thk123 replied:
Would it not be sufficent to have a:

```java
public class simpleClaz {
  String getStr() { return "hello"; }
}
/////////////
public class testMaybeNull1
{
    public simpleClaz x;
    public String check()
    {
        return x.getStr();
    }
}
```
Since x may or may not be null isn't it the same?

romainbrenguier replied:
No this example is already handled correctly. It is the `@class_identifier` field that is a problem, and as far as I now it can only be access via `instanceof` and `getClass`. And I think `getClass` is the only function which generate a `cprover_string_literal` with something else than a constant.

peterschrammel replied:
> But it is difficule to test on CBMC because we don't have a model for Class and the main affected method is Object.getClass.

Getting 'java core models' into the cbmc repo would help. @cesaro has started some work on that.

thk123 reviewed (commented):
Some suggestions for reducing code duplication, but I think the important thing is there should be some tests.

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -336,15 +340,37 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
+  if(type == string_typet())
+    return true;
+
+  if(type.id() == ID_struct || type.id() == ID_union)
```
This comment got hidden but I think still applies:

Can you pull the common code between this method and `has_char_pointer_subtype` into a separate method (`has_type_have_subtype(const typet & type, const namespacet &ns, const typet &required_subtype)`)

romainbrenguier committed:
Extend symbol resolution to string_typet

We take into account expressions of type string_typet in the
symbol_resolve object, so that these values can be propagated.
This can be particularly usefull for getting class identifiers as
strings, since these are represented as expression type string_typet in
the goto program.

This also adds a has_subtype function with predicate as arg to generalize what has_char_pointer_subtype and
has_string_subtype are doing.

romainbrenguier committed:
Refactor has_char_pointer_subtype with has_subtype

We reuse has_subtype to simplify the function.
Also get rid of the namespace argument that was not used.

romainbrenguier committed:
Add unit test for has_subtype

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -311,33 +311,81 @@ bool is_char_pointer_type(const typet &type)
+  if(pred(type))
     return true;
 
+  // clang-format off
```
I think we have been so far using clang-formats style where it disagrees with the linter (And using `// NOLINTNEXTLINE`)

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -311,33 +311,81 @@ bool is_char_pointer_type(const typet &type)
+///         for example, the subtypes of a struct are the types of its
+///         components, the subtype of a pointer is the type it points to,
+///         etc...
+///         For instance in the type `t` defined by
```
Thanks :+1:

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -311,33 +311,81 @@ bool is_char_pointer_type(const typet &type)
+  // clang-format off
+  return has_subtype(
+    type,
+    [](const typet &subtype)
```
Think you can just pass in `is_char_pointer_type` rather than wrapping in predicate

```C++
return has_subtype(type, is_char_pointer_type);
```

thk123 reviewed (approved):
Thanks - looks good, just a couple of very small nits but no need for re-review

romainbrenguier committed:
Adding only needed equations in symbol resolution

For cprover strings, adding all the equations to the union_find_replace
structure could sometimes become counter performant.
Instead here we look at what string expression are need for the function
calls and in which equations they appear.

romainbrenguier committed:
Unit test for string symbol resolution

romainbrenguier committed:
Debug information for string equations

We number equations to make it easier to debug function manipulating
equations, in particular string_identifiers_resolution_from_equations.
We also output symbol resolution information for strings

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -479,12 +481,118 @@ static union_find_replacet generate_symbol_resolution_from_equations(
+class equation_symbol_mappingt
+{
+public:
+  // Record index of the equations that contain a given expressions
```
expression

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -479,12 +481,118 @@ static union_find_replacet generate_symbol_resolution_from_equations(
+}
+
+/// Symbol resolution for expressions of type string typet.
+/// We record equality between this expressions in the output if one of the
```
these

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -479,12 +481,118 @@ static union_find_replacet generate_symbol_resolution_from_equations(
+
+/// Symbol resolution for expressions of type string typet.
+/// We record equality between this expressions in the output if one of the
+/// function calls depend on them.
```
depends

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -494,35 +602,59 @@ static union_find_replacet string_identifiers_resolution_from_equations(
     }
   }
+
+  // transitively add all equation which depends on the equations to treat
```
equations depend

peterschrammel commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -447,14 +481,191 @@ static union_find_replacet generate_symbol_resolution_from_equations(
+class equation_symbol_mappingt
+{
+public:
+  // Record index of the equations that contain a given expressions
```
expression

peterschrammel commented on src/solvers/refinement/string_refinement.h:
```diff
@@ -116,4 +116,9 @@ bool has_subtype(
   const typet &type,
   const std::function<bool(const typet &)> &pred);
 
+union_find_replacet string_identifiers_resolution_from_equations(
```
Add a comment that this is required to unit-test this function.

romainbrenguier commented:
@thk123 @mgudemann @smowton the version you reviewed was causing some performance problem on some tests (on a private repository), so I improved the code to only consider string equations that are actually used in a function. Can you have a look at the 3 last commits?

@peterschrammel can you have a look at this PR?

smowton commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -479,12 +481,118 @@ static union_find_replacet generate_symbol_resolution_from_equations(
+    strings_in_equation[i].push_back(expr);
+  }
+
+  std::vector<exprt> find_expressions(const std::size_t &i)
```
Little point taking a value type by const ref

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -479,12 +481,118 @@ static union_find_replacet generate_symbol_resolution_from_equations(
 }
 
-/// Symbol resolution for expressions of type string typet
+class equation_symbol_mappingt
```
Be great to add a doxygen summary of what this mapping is for

smowton commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -479,12 +481,118 @@ static union_find_replacet generate_symbol_resolution_from_equations(
+    {
+      const std::vector<exprt> strings_in_comp = extract_strings_from_lhs(
+        member_exprt(lhs, comp.get_name(), comp.type()));
+      for(const auto s : strings_in_comp)
```
result.insert(result.end(), strings_in_comp.begin(), strings_in_comp.end())

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -479,12 +481,118 @@ static union_find_replacet generate_symbol_resolution_from_equations(
+  return result;
+}
+
+/// \param eq: equation to add
```
Missing summary comment

smowton reviewed (commented):
Couple of nitpicks from the new commits

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -494,35 +602,59 @@ static union_find_replacet string_identifiers_resolution_from_equations(
+  {
+    const std::size_t i = equations_to_treat.top();
+    equations_to_treat.pop();
+    for(const exprt &string : equation_map.find_expressions(i))
```
Nit since these for loops spread over multiple lines please add braces for clarity

thk123 commented on src/solvers/refinement/string_refinement.cpp:
```diff
@@ -620,6 +752,12 @@ decision_proceduret::resultt string_refinementt::dec_solve()
     string_identifiers_resolution_from_equations(equations, ns, debug());
+#ifdef DEBUG
+  debug() << "symbol resolve string:" << eom;
+  for(const auto &pair : string_id_symbol_resolve.to_vector())
```
Since body of loop over multiple lines please add braces for clarity

thk123 reviewed (commented):
Reviewed final 3 commits - looks reasonable
